yt_dlp/extractor/tudou.py

```diff
@@ -0,0 +1,80 @@
+from .common import InfoExtractor
+import time
+import re
+
+
+class TudouIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:play\.)?tudou\.com/v_show/(?P<id>id_[\w=]+).html[\w\W]+'
+    _TESTS = [{
+        'url': 'https://play.tudou.com/v_show/id_XNjAxNjI2OTU3Ng.html?',
+        'md5': 'to be updated',
+        'info_dict': {
+            'id': 'XNjAxNjI2OTU3Ng==',
+            'ext': 'mp4',
+            'title': '阿斯塔意识到哈里杀了人，自己被骗了',
+            'show_name': '外星居民 第一季',
+        },
+        # 'skip': 'testing skip function',
+    }]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+        data = self._search_json(r'window\.__INITIAL_DATA__\s*=', webpage, 'initial data', video_id)
+
+        # The .json file is much cleaner, thanks for the guide
+        video_id = data['data']['data']['data']['extra']['videoId']
+        title = data['data']['data']['data']['extra']['videoTitle']
+        show_name = data['data']['data']['data']['extra']['showName']
+
+        _, urlh = self._download_webpage_handle(
+            "https://log.mmstat.com/yt.gif", video_id, 'Retrieving cna info'
+        )
+        re_cna = re.compile(r'cna=([\w\W]+; expires)')
+        cna = re.findall(re_cna, urlh.headers['set-cookie'])[0].replace("; expires", "")
+        ts1 = time.time() * 1000
+
+        params1 = {
+            'vid': video_id,
+            'ccode': '050F',
+            'client_ip': '192.168.1.1',
+            'utid': cna,
+            'client_ts': ts1,
+            'ckey': '140#XvSoigEEzzPw6zo2K5XTwpN8s9xI9h7FQeSkq6lrpeQ3Qm8xyFnPG9Uxtq6aSBxXVXwkqv6L4/GKvKzgD5oqlbzxhQfIlJgkzFnb0OK7lpTzzPzbVXlqlbrofJH+V3hqabzi228++bP0EHmiuZsHofFc74upoC6MkpscNHvXzJK+//Wna4Zt+dTHiQWqckfMQdZWTBs1ZpU/wadIq8nYxy5uZ+cRepqZMra+XLkaqMgGBcF/Ie/igRJDCcHl4d28aId7B+XOW/V6+gNOtDc+y6piEy1V51R4rYd41m6FkoEE4ix4eQE3VY7wvREVNJfR54V4qc3aqV4zzx+dkH8STo4ABYgr27bP9Vi0NBwse6wOOAJfYbnYZQdqQ4rlte/TfegJDmiufcPHd7sUG/A0RYx3pHQ9wYa3lQS3MjIjWejRlGzQdt0fxqEhcHuu10zfo4lhobcXARy7rDT2fb6wJrCHf98c0l8iGiOTSXVZVvWiQMM1est+EACvaMFB/baJm1BjCMKeS2zQEgUkRmulz00icM/W0BS8sEOR1RpOP4WXpji1HjEDpt2MTVHGqUwo223u03IGTHK3Z+Ki1ujWXKKUP+E1VESmBF0rzBkR/dyP20hRbGSpfX7eRVABF6npnWgdgjzQB4FOxAjHPn5CupypFZDYIdKyfQyC45mGS5WI6Wz9JmoFjxO7ikUKKP6p0KP3nj+5YTtbwUNpciVhD+mOZZKq6+dT6G7cfBhOzZQFXs66gZ3174KWt6sLTcvLzEezfoA0mO0NIx0mZZlL1lYonycOtcByLpKmSW5Xd96FWP5fuenIBUUJZCectNPgB/nFNGpeZuhruYbBbG843gwtfiXaRp2r4ssDM07/pPsEfk6MzcmtAxIimwdRjv6lSx+QSEANWRazykZneHEpH1R0uB0BjJdozYDKK84KMsNd+jjg6XTz9Ai0KZYMUJXN1Vz=',
```

Comment:
Are you sure this isn't dynamically generated / won't expire / will work for other IP addresses than your own?

Reply:
ckey is generated by a formula which is comlex(at least to me...), a lot of code obfuscation in there.
But I do find ckey provided by others, like 2 years ago still work, so I opt to leave it be like this for now.